test/unittest/tests/test_futex.cpp

```diff
@@ -186,77 +186,88 @@ class futex_tester {
-                if (match->_type == WT_FUTEX_WAKE_ONE)
-                    wakes.erase(match);
+        if (find_if(wake_sigs.begin(), wake_sigs.end(), is_wake_all) != wake_sigs.end()) {
+            REQUIRE(wake_sigs.size() == 1);
+            auto sig = wake_sigs.front();
+            auto wake_cnt = count_if(wbeg, wend, bind(&waiter::awoken, _1, sig._value));
+            auto spurious_cnt = count_if(wbeg, wend, bind(&waiter::spurious, _1, sig._value));
+            REQUIRE(((wake_cnt + spurious_cnt) == _waiters.size()));
+            return (spurious_cnt > 0) ? outcome::SpuriousWakeups : outcome::AsExpected;
+        }
+
+        /* Account for any expected timeouts. */
+        list<waiter> rem_waiters{_waiters.begin(), _waiters.end()};
+        auto timeout_cnt = count_if(wbeg, wend, bind(&waiter::timedout, _1));
+        if (timeout_cnt > 0) {
+            /*
+             * The timeout count should match the difference in the number of explicit wake up
+             * signals and waiters.
+             */
+            if (timeout_cnt != (_waiters.size() - wake_sigs.size())) {
+                return (outcome::UnexpectedTimeouts);
             }
+            /* Now that timeouts are accounted for, they are no longer of interest. */
+            rem_waiters.remove_if(bind(&waiter::timedout, _1));
         }
 
-        if (timedout > timeouts) {
-            return UnexpectedTimeouts;
+        /* Match remaining explicit wake ups with remaining waiters. */
+        bool spurious_wakeups = false;
+        list<wake_signal> rem_sigs{wake_sigs.begin(), wake_sigs.end()};
+        while (!rem_sigs.empty() && !rem_waiters.empty()) {
+            auto sig = rem_sigs.front();
+            auto match = find_if(
+              rem_waiters.begin(), rem_waiters.end(), bind(&waiter::awoken, _1, sig._value));
+
+            if (match == rem_waiters.end()) {
+                /*
+                 * No matching waiter for the wake up, so there should be a waiter that awoke
+                 * spuriously.
+                 */
+                auto spurious = find_if(
+                  rem_waiters.begin(), rem_waiters.end(), bind(&waiter::spurious, _1, sig._value));
+                if (spurious == rem_waiters.end()) {
+                    return (outcome::LostWakeup);
+                }
+                rem_waiters.erase(spurious);
+                spurious_wakeups = true;
+            } else {
+                rem_waiters.erase(match);
+            }
+            rem_sigs.pop_front();
         }
 
-        if (wakes.empty() && timedout == timeouts)
-            return outcome::Expected;
+        /* Test validation check. */
+        REQUIRE((rem_waiters.empty() && rem_sigs.empty()));
 
-        /*
-         * Spurious wake ups detected: while this is not an error, is not a consider successful.
-         */
-        return outcome::Acceptable;
+        return ((spurious_wakeups) ? outcome::SpuriousWakeups : outcome::AsExpected);
```

Comment:
I like the test machinery refactor overall, but what's the value in having a separate `SpuriousWakeups` returned here? The one place that uses this utility function doesn't care.

Reply:
This is left over from when I thought it might be necessary to wrap the tests in a loop to try for an `AsExpected` result. I'm not eager to add more complexity, but it is not really used now, and I'm not sure hwo to signal that a test did not pass in an "ideal" sense, in the face of a possible but not usual result.